torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -1467,6 +1467,10 @@ void Reducer::finalize_backward() {
     auto future_result = comm_hook_ == nullptr
         ? detail::parseCppCommHookResult(bucket.future_work->value())
         : comm_hook_->parseHookResult(bucket.future_work->value());
+    // Convert back to complex as the collective might have not conveted it.
+    if (bucket.gradients.is_complex() && !future_result.is_complex()) {
```

Comment:
I wonder if we should do this "silent" conversion for user-defined comm. hooks or just leave the conversion up to users to implement and raise an appropriate error if they do not.

i.e. what if there are complex gradients, but user's optimizer assumes non-complex and the optimizer performs some sort of conversion?

Reply:
Hmmm, I tried to make the comms hook undo it but IIRC, I ran into issues around it.
I wish I had documented what was the issue, let me try that and see how it does.